### DISCUSSION TITLE
Revisions and relational items inconsistency

### DISCUSSION THREAD

**Discussion by lubrst** [directus/directus#14098] [Archive]
Today I fetched the latest revision of an item via the API and noticed that the revision data contains old values for relational fields (Directus version 9.12.2).

For example, there are two collections A and B.
Collection A has a m2o relationship associated with Collection B, with the relational trigger "nullify".
Now when I delete an item in B, the relationship field in the corresponding item in A is set to null, but no new revision is created for the item in A and the revision data still contains the id of the deleted item. (addition: when I open the revision entry in the app, I get an error because the app tries to fetch the deleted entry)

For other relationship types, this behavior could also occur.

Is this behavior by design?

And a second possible inconsistency:

It seems that when a new item is created, only the request payload is stored in the revision and not the entire new object (e.g. with all default values).

I noticed this behavior when checking the revisions of m2a items.
For example, there are two collections A and B with a m2a relationship and junction table A_to_all.

Create "A" item request payload:
```
{
  many_to_all_field: [
    {
      collection: "B",
      item: { ... }
    }
  ]
}
```

Revision data of auto-generated A_to_all item:
````
{
    collection: "B",
    item: { ... },
    A_id: "<new id A item>"
}
````

Expected revision data:
````
{
  collection "B",
  item: "<new id of B item>",
  A_item: "<new id A item>"
}
````

Is this behavior also by design, because revision data field documentation says "Snapshot of the top-level item data."?

**Comment by rijkvanzanten** [comment]
> Collection A has a m2o relationship associated with Collection B, with the relational trigger "nullify".
> Now when I delete an item in B, the relationship field in the corresponding item in A is set to null, but no new revision is created for the item in A and the revision data still contains the id of the deleted item. (addition: when I open the revision entry in the app, I get an error because the app tries to fetch the deleted entry)

This is correct. The relational triggers are all handled on the database directly, and are therefore invisible to Directus. 

> It seems that when a new item is created, only the request payload is stored in the revision and not the entire new object (e.g. with all default values).

This is the case for the `delta`, not the `data`. `data` is a "fresh" read using `fields=*`

**Reply by lubrst** [reply]
> This is the case for the delta, not the data. data is a "fresh" read using fields=*

In `updateMany`  (ItemService) a fresh read is done.

https://github.com/directus/directus/blob/e6a893e46e929c4ea24bb4b39db758326579bd8f/api/src/services/items.ts#L510-L528

But in `createOne` `data` is filled with the payload, same as `delta`.

https://github.com/directus/directus/blob/e6a893e46e929c4ea24bb4b39db758326579bd8f/api/src/services/items.ts#L174-L180

Should a fresh read be used here as well (or the result of the database insert)?

**Reply by rijkvanzanten** [reply]
Hmm yeah, I believe the original thinking was that the submitted payload would be identical to the thing inserted, but that excludes database-generated content
